library/core/src/ptr/const_ptr.rs

```diff
@@ -409,6 +409,23 @@ impl<T: ?Sized> *const T {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // Precondition 1: the computed offset `count * size_of::<T>()` does not overflow `isize`.
+        if let Some(computed_offset) = count.checked_mul(core::mem::size_of::<T>() as isize) {
+            // Precondition 2: adding the computed offset to `self` does not cause overflow.
+            (self as isize).checked_add(computed_offset).is_some()
+        } else {
+            false
+        } &&
+        // Precondition 3: If `T` is a unit type (`size_of::<T>() == 0`), this check is unnecessary as it has no allocated memory.
+        // Otherwise, for non-unit types, `self` and `self.wrapping_offset(count)` should point to the same allocated object,
+        // restricting `count` to prevent crossing allocation boundaries.
+        ((core::mem::size_of::<T>() == 0) || (kani::mem::same_allocation(self, self.wrapping_offset(count))))
```

Comment:
I recommend the following changes:
1. Start with the check that `T` is a ZST, in which case the operation should be safe.
2. You could also use `map_or(false, |computed_offset| {})` to replace the `if let` logic to make it a bit cleaner (not a big deal).
3.  Please don't use `kani` specific APIs in a contract. If you do need a tool specific API, the attribute should be conditional on the tool being enabled. I believe `same_allocation` has been added to `ub_checks` module. If it hasn't feel free to add it.

Something like:

```suggestion
        (core::mem::size_of::<T>() == 0)
        || (count.checked_mul(core::mem::size_of::<T>() as isize).map_or(false, |computed_offset| {
              (self as isize).checked_add(computed_offset).is_some()
            })
          && core::ub_checks::same_allocation(self, self.wrapping_offset(count)))
```

Reply:
Please, make sure to address this comment on the existing contracts added from https://github.com/model-checking/verify-rust-std/pull/113 and https://github.com/model-checking/verify-rust-std/pull/154.

Reply:
cc. @xsxszab @stogaru @szlee118 @unseenmars @MayureshJoshi25

Reply:
Since this PR is for verifying `const` functions, would it be appropriate to update the code for `mut` functions as well?

Reply:
you can address on a separate PR.

Reply:
Sure, I’ll submit a new PR once current comments are resolved.

Reply:
PR #203 has been created to refactor `mut` contracts and proof for contracts.